rust/flatbuffers/src/array.rs

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright 2021 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+use std::marker::PhantomData;
+use std::fmt::{Debug, Formatter, Result};
+use std::mem::size_of;
+use std::slice::from_raw_parts;
+use crate::vector::{VectorIter, SafeSliceAccess};
+use crate::follow::Follow;
+
+#[derive(Copy, Clone)]
+pub struct Array<'a, T: 'a, const N: usize>(&'a [u8], PhantomData<T>);
+
+impl<'a, T: 'a, const N: usize> Debug for Array<'a, T, N>
+where
+    T: 'a + Follow<'a>,
+    <T as Follow<'a>>::Inner: Debug,
+{
+    fn fmt(&self, f: &mut Formatter) -> Result {
+        f.debug_list().entries(self.iter()).finish()
+    }
+}
+
+impl<'a, T: 'a, const N: usize> Array<'a, T, N> {
+    #[inline(always)]
+    pub fn new(buf: &'a [u8]) -> Self {
+        Array {
+            0: buf,
+            1: PhantomData,
+        }
+    }
+
+    #[inline(always)]
+    pub fn len(&self) -> usize {
+        N
+    }
+}
+
+impl<'a, T: Follow<'a> + 'a, const N: usize> Array<'a, T, N> {
+    #[inline(always)]
+    pub fn get(&self, idx: usize) -> T::Inner {
+        debug_assert!(idx < N);
+        let sz = size_of::<T>();
+        T::follow(self.0, sz * idx)
+    }
+
+    #[inline(always)]
+    pub fn iter(&self) -> VectorIter<'a, T> {
+        VectorIter::from_slice(self.0, self.len())
+    }
+}
+
+impl<'a, T: Follow<'a> + Debug, const N: usize> Into<[T::Inner; N]> for Array<'a, T, N>{
+    fn into(self) -> [T::Inner; N] {
+        let mut mem = core::mem::MaybeUninit::<[T::Inner; N]>::uninit();
+        let sz = core::mem::size_of::<T::Inner>();
+        let mem_ptr = mem.as_mut_ptr() as *mut u8;
+        unsafe {
+            for (i, item) in self.iter().enumerate() {
+                core::ptr::copy_nonoverlapping(
+                &item as *const T::Inner as *const u8,
+                mem_ptr.offset((i * sz) as isize),
+                sz,
+                );
+            }
+            mem.assume_init()
+        }
+    }
+}
+
+impl<'a, T: SafeSliceAccess + 'a, const N: usize> Array<'a, T, N> {
+    pub fn safe_slice(self) -> &'a [T] {
+        let sz = size_of::<T>();
+        debug_assert!(sz > 0);
+        let ptr = self.0.as_ptr() as *const T;
+        unsafe { from_raw_parts(ptr, N) }
```

Comment:
does this work? I think Array has alignement 1 while T may have higher alignment

Reply:
It is. Just like in the vector case on insert, everything is being aligned to 8 bytes on the creation of the struct.

Reply:
> It is. Just like in the vector case on insert, everything is being aligned to 8 bytes on the creation of the struct.

Can you rephrase that? I don't understand what you mean.

---


Also, I guess this code is fine since `T: SafeSliceAccess`. 
That means `T` is alignment-1 and endian agnostic.
Basically `T` is `i8, u8`, or a flatbuffers-struct.
(`SafeSliceAccess` really should be an `unsafe trait`).

Note that Flatbuffers reads data coming from a `&[u8]` so we cannot assume things are aligned to 8 bytes in the computer's memory, only that its aligned to 8 bytes w.r.t the buffer we're reading from.

Reply:
From my understanding it is critical to be aligned with the buffer we are reading from. So as long as we have the same alignment and padding in the structs as the bytes we read everything will be fine. The memory alignment is more for cpu performance so it will be able to fetch ints in 1 instruction and won't have to fetch more than once because the number is not aligned. Am I missing something? 🤔

Reply:
>  So as long as we have the same alignment and padding in the structs as the bytes we read everything will be fine.

We verify that types within the buffer are aligned wrt the buffer but the problem is that there's no guarantee that the buffer is aligned wrt the computer's actual memory. `get_root` takes a `&[u8]` and unlike C++'s `std::vector`, the first address in the slice's buffer is not guaranteed to be alignment 16. jemalloc actually takes advantage of this and will give unaligned `Vec<u8>` for small data. See #6359. So when you do `from_raw_parts(ptr, N)`, `ptr` may not be aligned with the computer's memory.

> The memory alignment is more for cpu performance so it will be able to fetch ints in 1 instruction and won't have to fetch more than once because the number is not aligned. 

Actually unaligned memory access is UB. It turns out to be just slower on x86 but it'll do BadThings on other architectures like ARM.

Anyway, I'll resolve this conversation since `T: SafeSliceAccess` so none of the alignment errors will show up.


---
Extra historical notes:

We mostly use SafeSliceAccess so we can sometimes return `&'a [T]` instead of `flatbuffers::Vector<'a, T>`. Imo that was a bad design choice for a couple reasons:
* It introduces complexity in code-gen with this SafeSliceAccess stuff
* the API is inconsistent with `&'a[T]` being used when `T` is `i8,u8` or a struct.

We implement `SafeSliceAccess` for >1byte primitives (e.g. `f32,u16,i64`) on little-endian targets. That's not used in the generated-code but I think the original implementer wanted to use `&'a[T]` instead of `flatbuffers::Vector<'a, T>` for those primitives on little-endian targets too. However, that would've forced clients to conditionally compile code to handle our conditional types, which is _even more_ complexity, all so we can avoid using `flatbuffers::Vector<'a, T>`.

Overall, I'm very critical of the existence of SafeSliceAccess.